xdem/volume.py

```diff
@@ -506,3 +507,194 @@ def local_hypsometric_interpolation(voided_ddem: Union[np.ndarray, np.ma.masked_
     assert output is not None
 
     return output
+
+
+def get_regional_hypsometric_signal(ddem: Union[np.ndarray, np.ma.masked_array],
+                                    ref_dem: Union[np.ndarray, np.ma.masked_array],
+                                    glacier_index_map: np.ndarray) -> pd.DataFrame:
+    """
+    Get the normalized regional hypsometric elevation change signal, read "the general shape of it".
+
+    :param ddem: The dDEM to analyse.
+    :param ref_dem: A void-free reference DEM.
+    :param glacier_index_map: An array glacier indices of the same shape as the previous inputs.
+    """
+    n_bins = 20  # TODO: This should be an argument.
+
+    # Extract the array and mask representations of the arrays.
+    ddem_arr, ddem_mask = xdem.spatial_tools.get_array_and_mask(ddem.squeeze())
+    ref_arr, ref_mask = xdem.spatial_tools.get_array_and_mask(ref_dem.squeeze())
+
+    # The reference DEM should be void free
+    assert np.count_nonzero(ref_mask) == 0, "Reference DEM has voids"
+
+    # The unique indices are the unique glaciers.
+    unique_indices = np.unique(glacier_index_map)
+
+    # Generate bins of normalized elevation.
+    elev_norm = np.linspace(0, 1, n_bins)
+
+    # Create empty (ddem) value and (pixel) count arrays which will be filled iteratively.
+    values = np.empty((n_bins, unique_indices.shape[0]), dtype=float) + np.nan
+    counts = np.empty((n_bins, unique_indices.shape[0]), dtype=float) + np.nan
+
+    # Start a counter of glaciers that are actually processed.
+    count = 0
+    # Loop over each unique glacier.
+    for i in np.unique(glacier_index_map):
+        # If i ==0, it's assumed to be periglacial.
+        if i == 0:
+            continue
+        # Create a mask representing a particular glacier.
+        glacier_values = (glacier_index_map == i)
+
+        # Stop if the amount of pixels is tiny.
+        # TODO: Make this more dynamic.
+        if np.count_nonzero(glacier_values) < 10:
+            continue
+        # At this point, invalid glaciers have been skipped and the counter should be incremented.
+        count += 1
+
+        # The inlier mask is where that particular glacier is and where nans don't exist.
+        inlier_mask = glacier_values & ~ddem_mask
+
+        # Extract only the difference and elevation values that correspond to the glacier.
+        differences = ddem_arr[inlier_mask]
+        elevations = ref_arr[inlier_mask]
+
+        # Run the hypsometric binning.
+        bins = hypsometric_binning(differences, elevations, bins=n_bins, kind="count")
+
+        # Normalize by elevation.
+        bins.index = bins.index.mid
+        bins.index -= bins.index.min()
+        bins.index /= bins.index.max()
```

Comment:
It's strange that minmax scaling isn't a builtin numpy function! Also, I just read that normalization includes division by the total sum, whereas min-max scaling just scales the values between the minimum and maximum (like above). 

Because of this, (assuming min-max scaling is the one we want), shouldn't the function be called "Scaled regional hypsometric interpolation" instead of "normalized ..."? What do you think, @rhugonnet and @adehecq?

Reply:
Added neater scaling in 3e53f45